File: packages/shared/src/queries.ts
import type { SupabaseClient } from "@supabase/supabase-js";
import type {
  BookingRequest,
  Category,
  Conversation,
  Listing,
  ListingWithJoins,
  Message,
  Profile,
} from "./types";

const LISTING_SELECT =
  "*, provider:profiles(*), category:categories(*)";

export async function fetchCategories(db: SupabaseClient): Promise<Category[]> {
  const { data, error } = await db.from("categories").select("*").order("id");
  if (error) throw error;
  return data as Category[];
}

export interface BrowseFilters {
  categoryId?: number;
  query?: string;
  lat?: number;
  lng?: number;
  radiusKm?: number;
}

export async function fetchListings(
  db: SupabaseClient,
  filters: BrowseFilters = {}
): Promise<ListingWithJoins[]> {
  // Use geo RPC when a location is set, otherwise a plain filtered query.
  if (filters.lat != null && filters.lng != null) {
    const { data, error } = await db.rpc("listings_nearby", {
      p_lat: filters.lat,
      p_lng: filters.lng,
      p_radius_km: filters.radiusKm ?? 40,
      p_category: filters.categoryId ?? null,
      p_query: filters.query || null,
    });
    if (error) throw error;
    return hydrateListings(db, data as Listing[]);
  }

  let q = db
    .from("listings")
    .select(LISTING_SELECT)
    .eq("status", "active")
    .order("created_at", { ascending: false })
    .limit(60);
  if (filters.categoryId) q = q.eq("category_id", filters.categoryId);
  if (filters.query)
    q = q.or(`title.ilike.%${filters.query}%,description.ilike.%${filters.query}%`);
  const { data, error } = await q;
  if (error) throw error;
  return data as unknown as ListingWithJoins[];
}

async function hydrateListings(
  db: SupabaseClient,
  listings: Listing[]
): Promise<ListingWithJoins[]> {
  if (!listings.length) return [];
  const ids = listings.map((l) => l.id);
  const { data, error } = await db
    .from("listings")
    .select(LISTING_SELECT)
    .in("id", ids);
  if (error) throw error;
  const byId = new Map((data as unknown as ListingWithJoins[]).map((l) => [l.id, l]));
  return listings.map((l) => byId.get(l.id) ?? l);
}

export async function fetchListing(
  db: SupabaseClient,
  id: string
): Promise<ListingWithJoins | null> {
  const { data, error } = await db
    .from("listings")
    .select(LISTING_SELECT)
    .eq("id", id)
    .maybeSingle();
  if (error) throw error;
  return data as unknown as ListingWithJoins | null;
}

export async function fetchMyListings(
  db: SupabaseClient,
  providerId: string
): Promise<ListingWithJoins[]> {
  const { data, error } = await db
    .from("listings")
    .select(LISTING_SELECT)
    .eq("provider_id", providerId)
    .neq("status", "deleted")
    .order("created_at", { ascending: false });
  if (error) throw error;
  return data as unknown as ListingWithJoins[];
}

export async function getOrCreateConversation(
  db: SupabaseClient,
  listing: Listing,
  customerId: string
): Promise<Conversation> {
  const { data: existing } = await db
    .from("conversations")
    .select("*")
    .eq("listing_id", listing.id)
    .eq("customer_id", customerId)
    .maybeSingle();
  if (existing) return existing as Conversation;

  const { data, error } = await db
    .from("conversations")
    .insert({
      listing_id: listing.id,
      customer_id: customerId,
      provider_id: listing.provider_id,
    })
    .select("*")
    .single();
  if (error) throw error;
  return data as Conversation;
}

export async function fetchConversations(
  db: SupabaseClient,
  userId: string
): Promise<Conversation[]> {
  const { data, error } = await db
    .from("conversations")
    .select("*, listing:listings(*), customer:profiles!conversations_customer_id_fkey(*), provider:profiles!conversations_provider_id_fkey(*)")
    .or(`customer_id.eq.${userId},provider_id.eq.${userId}`)
    .order("created_at", { ascending: false });
  if (error) throw error;
  return data as unknown as Conversation[];
}

export async function fetchMessages(
  db: SupabaseClient,
  conversationId: string
): Promise<Message[]> {
  const { data, error } = await db
    .from("messages")
    .select("*")
    .eq("conversation_id", conversationId)
    .order("created_at");
  if (error) throw error;
  return data as Message[];
}

export async function sendMessage(
  db: SupabaseClient,
  conversationId: string,
  senderId: string,
  body: string
): Promise<Message> {
  const { data, error } = await db
    .from("messages")
    .insert({ conversation_id: conversationId, sender_id: senderId, body })
    .select("*")
    .single();
  if (error) throw error;
  return data as Message;
}

export async function createBookingRequest(
  db: SupabaseClient,
  input: {
    listing: Listing;
    customerId: string;
    requestedAt: string;
    address?: string;
    details?: string;
  }
): Promise<BookingRequest> {
  const { data, error } = await db
    .from("booking_requests")
    .insert({
      listing_id: input.listing.id,
      customer_id: input.customerId,
      provider_id: input.listing.provider_id,
      requested_at: input.requestedAt,
      address: input.address ?? null,
      details: input.details ?? "",
    })
    .select("*")
    .single();
  if (error) throw error;
  return data as BookingRequest;
}

export async function fetchBookings(
  db: SupabaseClient,
  userId: string
): Promise<BookingRequest[]> {
  const { data, error } = await db
    .from("booking_requests")
    .select("*, listing:listings(*), customer:profiles!booking_requests_customer_id_fkey(*), provider:profiles!booking_requests_provider_id_fkey(*)")
    .or(`customer_id.eq.${userId},provider_id.eq.${userId}`)
    .order("created_at", { ascending: false });
  if (error) throw error;
  return data as unknown as BookingRequest[];
}

export async function updateBookingStatus(
  db: SupabaseClient,
  id: string,
  status: BookingRequest["status"]
): Promise<void> {
  const { error } = await db.from("booking_requests").update({ status }).eq("id", id);
  if (error) throw error;
}

export async function fetchProfile(
  db: SupabaseClient,
  id: string
): Promise<Profile | null> {
  const { data, error } = await db
    .from("profiles")
    .select("*")
    .eq("id", id)
    .maybeSingle();
  if (error) throw error;
  return data as Profile | null;
}

File: apps/streamlit/lib/queries.py
"""Data-access helpers mirroring packages/shared's queries.ts, so this demo
reads/writes the exact same Supabase schema as the web and mobile apps."""

from datetime import datetime, timezone

LISTING_SELECT = "*, provider:profiles(*), category:categories(*)"


def fetch_categories(db):
    res = db.table("categories").select("*").order("id").execute()
    return res.data


def fetch_listings(db, category_id=None, query=None, lat=None, lng=None, radius_km=40):
    if lat is not None and lng is not None:
        res = db.rpc(
            "listings_nearby",
            {
                "p_lat": lat,
                "p_lng": lng,
                "p_radius_km": radius_km,
                "p_category": category_id,
                "p_query": query or None,
            },
        ).execute()
        listings = res.data
        if not listings:
            return []
        ids = [l["id"] for l in listings]
        hydrated = (
            db.table("listings").select(LISTING_SELECT).in_("id", ids).execute().data
        )
        by_id = {l["id"]: l for l in hydrated}
        return [by_id.get(l["id"], l) for l in listings]

    q = (
        db.table("listings")
        .select(LISTING_SELECT)
        .eq("status", "active")
        .order("created_at", desc=True)
        .limit(60)
    )
    if category_id:
        q = q.eq("category_id", category_id)
    if query:
        q = q.or_(f"title.ilike.%{query}%,description.ilike.%{query}%")
    return q.execute().data


def fetch_listing(db, listing_id):
    res = (
        db.table("listings")
        .select(LISTING_SELECT)
        .eq("id", listing_id)
        .maybe_single()
        .execute()
    )
    return res.data if res else None


def fetch_my_listings(db, provider_id):
    res = (
        db.table("listings")
        .select(LISTING_SELECT)
        .eq("provider_id", provider_id)
        .neq("status", "deleted")
        .order("created_at", desc=True)
        .execute()
    )
    return res.data


def create_listing(db, provider_id, **fields):
    payload = {"provider_id": provider_id, **fields}
    inserted = db.table("listings").insert(payload).execute().data[0]
    db.table("profiles").update({"is_provider": True}).eq("id", provider_id).execute()
    return fetch_listing(db, inserted["id"])


def set_listing_status(db, listing_id, status):
    db.table("listings").update({"status": status}).eq("id", listing_id).execute()


def get_or_create_conversation(db, listing, customer_id):
    existing = (
        db.table("conversations")
        .select("*")
        .eq("listing_id", listing["id"])
        .eq("customer_id", customer_id)
        .maybe_single()
        .execute()
    )
    if existing and existing.data:
        return existing.data
    res = (
        db.table("conversations")
        .insert(
            {
                "listing_id": listing["id"],
                "customer_id": customer_id,
                "provider_id": listing["provider_id"],
            }
        )
        .execute()
    )
    return res.data[0]


def fetch_conversations(db, user_id):
    res = (
        db.table("conversations")
        .select(
            "*, listing:listings(*), "
            "customer:profiles!conversations_customer_id_fkey(*), "
            "provider:profiles!conversations_provider_id_fkey(*)"
        )
        .or_(f"customer_id.eq.{user_id},provider_id.eq.{user_id}")
        .order("created_at", desc=True)
        .execute()
    )
    return res.data


def fetch_messages(db, conversation_id):
    res = (
        db.table("messages")
        .select("*")
        .eq("conversation_id", conversation_id)
        .order("created_at")
        .execute()
    )
    return res.data


def send_message(db, conversation_id, sender_id, body):
    res = (
        db.table("messages")
        .insert({"conversation_id": conversation_id, "sender_id": sender_id, "body": body})
        .execute()
    )
    return res.data[0]


def create_booking_request(db, listing, customer_id, requested_at, address=None, details=""):
    res = (
        db.table("booking_requests")
        .insert(
            {
                "listing_id": listing["id"],
                "customer_id": customer_id,
                "provider_id": listing["provider_id"],
                "requested_at": requested_at,
                "address": address,
                "details": details or "",
            }
        )
        .execute()
    )
    return res.data[0]


def fetch_bookings(db, user_id):
    res = (
        db.table("booking_requests")
        .select(
            "*, listing:listings(*), "
            "customer:profiles!booking_requests_customer_id_fkey(*), "
            "provider:profiles!booking_requests_provider_id_fkey(*)"
        )
        .or_(f"customer_id.eq.{user_id},provider_id.eq.{user_id}")
        .order("created_at", desc=True)
        .execute()
    )
    return res.data


def update_booking_status(db, booking_id, status):
    db.table("booking_requests").update({"status": status}).eq("id", booking_id).execute()


def update_profile(db, user_id, **fields):
    db.table("profiles").update(fields).eq("id", user_id).execute()


# ---------- formatting helpers (mirrors packages/shared/src/format.ts) ----------

def price_label(listing):
    unit = listing.get("price_unit")
    price = listing.get("price")
    if unit == "quote" or price is None:
        return "Get a quote"
    amount = f"${float(price):.0f}" if float(price) % 1 == 0 else f"${float(price):.2f}"
    return f"{amount}/hr" if unit == "hourly" else amount


def time_ago(iso):
    if not iso:
        return ""
    then = datetime.fromisoformat(iso.replace("Z", "+00:00"))
    seconds = (datetime.now(timezone.utc) - then).total_seconds()
    if seconds < 60:
        return "just now"
    minutes = int(seconds // 60)
    if minutes < 60:
        return f"{minutes}m ago"
    hours = int(minutes // 60)
    if hours < 24:
        return f"{hours}h ago"
    days = int(hours // 24)
    if days < 7:
        return f"{days}d ago"
    return then.strftime("%b %-d, %Y")
